### PR TITLE
refactor(theme-search-algolia): migrate package to TS

### PR DESCRIPTION
**Pull request by armano2**
## Motivation

#5817

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests)?

yes

### CODE REVIEW

File: packages/docusaurus-theme-search-algolia/src/theme/hooks/useAlgoliaContextualFacetFilters.ts [hunk 1]
```diff
@@ -6,9 +6,10 @@
  */
 
 import useContextualSearchFilters from '@theme/hooks/useContextualSearchFilters';
```
Comment (armano2) [inline packages/docusaurus-theme-search-algolia/src/theme/hooks/useAlgoliaContextualFacetFilters.ts:8]: i'm unsure if this is correct, we are importing this hook from theme-classic and thus compilation doesn't work as this package doesn't rely on theme

**Comment by armano2** [comment]
There is small issue with migrating this package, `useContextualSearchFilters` is defined within `theme-classic` and `theme-search-algolia` doesn't rely and see this package

to fix this we have to do one of 2 thinks
- `theme-search-algolia` has dependency on `theme-classic`
- `useContextualSearchFilters` is moved to `theme-common`

i can fix it both ways, but I'm unsure witch one is preferred @Josh-Cena

File: packages/docusaurus-theme-search-algolia/package.json [hunk 2]
```diff
@@ -12,6 +13,12 @@
     "directory": "packages/docusaurus-theme-search-algolia"
   },
   "license": "MIT",
+  "scripts": {
+    "build": "yarn build:server & yarn build:browser & yarn build:copy",
+    "build:server": "tsc --project tsconfig.server.json",
+    "build:browser": "tsc --project tsconfig.browser.json",
+    "build:copy": "node copyUntypedFiles.js"
+  },
```
Comment (armano2) [inline packages/docusaurus-theme-search-algolia/package.json:21]: not sure if this is fine, there is no current example of package with build of ts/x files that support both server and browser code

File: packages/docusaurus-theme-search-algolia/package.json [hunk 3]
```diff
@@ -12,6 +13,13 @@
     "directory": "packages/docusaurus-theme-search-algolia"
   },
   "license": "MIT",
+  "scripts": {
+    "build": "yarn build:server & yarn build:browser & yarn build:next & yarn build:copy",
+    "build:server": "tsc --project tsconfig.server.json",
+    "build:browser": "tsc --project tsconfig.browser.json",
+    "build:next": "tsc --project tsconfig.browser.json --jsx preserve --outDir lib-next",
+    "build:copy": "node copyUntypedFiles.js"
+  },
```
Comment (armano2) [inline packages/docusaurus-theme-search-algolia/package.json:22]: not sure if this is fine, there is no current example of package with build of ts/x files that support both server and browser code

Comment (armano2) [reply packages/docusaurus-theme-search-algolia/package.json:22]: next build and browser should be unified, i'm just unsure why *.jsx files are not correctly resolved by webpack

do you have idea?

Comment (Josh-Cena) [reply packages/docusaurus-theme-search-algolia/package.json:22]: You can output `.js` extension with `"jsx": "react-native"`

Comment (armano2) [reply packages/docusaurus-theme-search-algolia/package.json]: ok updated

**Comment by slorber** [comment]
Thanks 👍 good enough to merge

Now that we have TS here it will be better to refactor a bit because it's quite messy 🤪

**Pull request merged**